rules/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictParamRector.php

```diff
@@ -0,0 +1,219 @@
+use PhpParser\Node\Param;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\Node\Stmt\Function_;
+use PhpParser\Node\Stmt\Return_;
+use PhpParser\NodeTraverser;
+use PHPStan\Analyser\Scope;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\ObjectType;
+use PHPStan\Type\TypeCombinator;
+use PHPStan\Type\UnionType;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\Rector\AbstractScopeAwareRector;
+use Rector\Core\ValueObject\PhpVersionFeature;
+use Rector\TypeDeclaration\TypeInferer\ReturnTypeInferer;
+use Rector\VendorLocker\ParentClassMethodTypeOverrideGuard;
+use Rector\VersionBonding\Contract\MinPhpVersionInterface;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\TypeDeclaration\Rector\ClassMethod\ReturnTypeFromStrictParamRector\ReturnTypeFromStrictParamRectorTest
+ */
+final class ReturnTypeFromStrictParamRector extends AbstractScopeAwareRector implements MinPhpVersionInterface
+{
+    public function __construct(
+        private readonly ParentClassMethodTypeOverrideGuard $parentClassMethodTypeOverrideGuard,
+        private readonly ReturnTypeInferer $returnTypeInferer
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Add return type based on strict parameter type', [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public function resolve(ParamType $item)
+    {
+        return $item;
+    }
+}
+CODE_SAMPLE
+
+                ,
+                <<<'CODE_SAMPLE'
+class SomeClass
+{
+    public function resolve(ParamType $item): ParamType
+    {
+        return $item;
+    }
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [ClassMethod::class, Function_::class];
+    }
+
+    public function provideMinPhpVersion(): int
+    {
+        return PhpVersionFeature::NULLABLE_TYPE;
+    }
+
+    /**
+     * @param ClassMethod|Function_ $node
+     */
+    public function refactorWithScope(Node $node, Scope $scope): ?Node
+    {
+        if ($this->shouldSkipNode($node)) {
+            return null;
+        }
+
+        $return = $this->findCurrentScopeReturn($node);
+        if ($return === null || $return->expr === null) {
+            return null;
+        }
+
+        $returnName = $this->getName($return->expr);
+        foreach ($node->getParams() as $param) {
+            if (!$param->type instanceof Node) {
+                continue;
+            }
+
+            if ($this->shouldSkipParam($param, $node)) {
+                continue;
+            }
+
+            $paramName = $this->getName($param);
+            if ($returnName !== $paramName) {
+                continue;
+            }
+
+            $node->returnType = $param->type;
+            return $node;
```

Comment:
what about multiple params?

```php
function foo(int $a, bool $b) {
    if (rand(0, 1)) {
       return $a;
    }

   return $b;
}
```

Reply:
covered in rules-tests/TypeDeclaration/Rector/ClassMethod/ReturnTypeFromStrictParamRector/Fixture/skip_multiple_returns.php.inc